src/main/java/com/github/manolo8/darkbot/core/objects/ShipInfo.java

```diff
@@ -24,6 +30,33 @@ public void update() {
 
         destination.update(API.readMemoryLong(address + 96));
         destination.update();
+
+        calcSpeed();
+    }
+
+    private double averageSpeed, pastTimeNeeded;
+    private Location pastDestination;
+
+    private void calcSpeed() {
+        if (speed > 50) return;
+
+        double timeNeeded = API.readDouble(address, 104, 152);
+        if (destination.now.equals(pastDestination) && pastTimeNeeded == timeNeeded) return;
+
+        this.pastDestination = destination.now.copy();
+        this.pastTimeNeeded = timeNeeded;
+
+        double elapsed = API.readDouble(address, 104, 136);
+
+        double newSpeed = currentLoc.distance(destination) / (timeNeeded - elapsed);
+        if (newSpeed < 100) return;
+        if (newSpeed > 1000) newSpeed = 1000;
+
+        if (averageSpeed == 0) averageSpeed = newSpeed;
+        else averageSpeed = averageSpeed * 0.75 + newSpeed * 0.25;
```

Comment:
```suggestion
        if (predictedSpeed == 0) predictedSpeed = newSpeed;
        else predictedSpeed = predictedSpeed * 0.75 + newSpeed * 0.25;
```
I think predictedSpeed is a better name for this variable